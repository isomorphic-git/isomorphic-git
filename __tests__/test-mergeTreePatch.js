/* eslint-env node, browser, jasmine */
const { makeFixture } = require('./__helpers__/FixtureFS.js')
const snapshots = require('./__snapshots__/test-diffTree.js.snap')
const registerSnapshots = require('./__helpers__/jasmine-snapshots')

const { mergeTreePatches } = require('isomorphic-git/internal-apis')
// @ts-ignore
const { E, diffTree } = require('isomorphic-git')

describe('mergeTreePatches', () => {
  beforeAll(() => {
    registerSnapshots(snapshots)
  })

  it('merge two empty patchsets', async () => {
    let patchset = await mergeTreePatches({ patchsets: [[], []]})
    expect(patchset).toEqual([])
  })

  it('merge two conflicting patchsets', async () => {
    let error = null
    try {
      await mergeTreePatches({ patchsets: [
        [{ filepath: 'a' }],
        [{ filepath: 'a' }]
      ]})
    } catch (e) {
      error = e
    }
    expect(error).not.toBeNull()
    expect(error.code).toBe(E.MergeConflict)
  })

  it('merge add-files and change-modes', async () => {
    // Setup
    let { gitdir } = await makeFixture('test-diff')
    // Test
    let diff1 = await diffTree({ gitdir, before: 'mainline', after: 'add-files' })
    let diff2 = await diffTree({ gitdir, before: 'mainline', after: 'change-modes' })
    let patchset = await mergeTreePatches({ patchsets: [diff1, diff2]})
    expect(patchset).toMatchSnapshot()
  })
  
  it('merge add-files and remove-files', async () => {
    // Setup
    let { gitdir } = await makeFixture('test-diff')
    // Test
    let diff1 = await diffTree({ gitdir, before: 'mainline', after: 'add-files' })
    let diff2 = await diffTree({ gitdir, before: 'mainline', after: 'remofe-files' })
    let patchset = await mergeTreePatches({ patchsets: [diff1, diff2]})
    expect(patchset).toMatchSnapshot()
  })

})
