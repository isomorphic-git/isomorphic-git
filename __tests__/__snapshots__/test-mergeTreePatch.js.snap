// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeTreePatches merge add-files and change-modes 1`] = `
Array [
  Object {
    "after": "01058d844a98d293a3b03a8615a34700e4ed2be3",
    "before": undefined,
    "filepath": "c/g",
    "ops": Array [
      "write",
    ],
  },
  Object {
    "after": "6a69f92020f5df77af6e8813ff1232493383b708",
    "before": undefined,
    "filepath": "f",
    "ops": Array [
      "write",
    ],
  },
  Object {
    "after": "78981922613b2afb6025042ff6bd878ac1994e85",
    "before": "78981922613b2afb6025042ff6bd878ac1994e85",
    "filepath": "a",
    "ops": Array [
      "chmod+x",
    ],
  },
  Object {
    "after": "4bcfe98e640c8284511312660fb8709b0afa888e",
    "before": "4bcfe98e640c8284511312660fb8709b0afa888e",
    "filepath": "c/d",
    "ops": Array [
      "chmod+x",
    ],
  },
]
`;

exports[`mergeTreePatches merge add-files and remove-files 1`] = `
Array [
  Object {
    "after": "01058d844a98d293a3b03a8615a34700e4ed2be3",
    "before": undefined,
    "filepath": "c/g",
    "ops": Array [
      "write",
    ],
  },
  Object {
    "after": "6a69f92020f5df77af6e8813ff1232493383b708",
    "before": undefined,
    "filepath": "f",
    "ops": Array [
      "write",
    ],
  },
  Object {
    "after": undefined,
    "before": "78981922613b2afb6025042ff6bd878ac1994e85",
    "filepath": "a",
    "ops": Array [
      "rm",
    ],
  },
  Object {
    "after": undefined,
    "before": "61780798228d17af2d34fce4cfbdf35556832472",
    "filepath": "b",
    "ops": Array [
      "rm",
    ],
  },
  Object {
    "after": undefined,
    "before": "4bcfe98e640c8284511312660fb8709b0afa888e",
    "filepath": "c/d",
    "ops": Array [
      "rm",
    ],
  },
  Object {
    "after": undefined,
    "before": "d905d9da82c97264ab6f4920e20242e088850ce9",
    "filepath": "c/e",
    "ops": Array [
      "rm",
    ],
  },
  Object {
    "after": undefined,
    "before": "faceb189bb665fe7a631602d904e57a4014187b7",
    "filepath": "c",
    "ops": Array [
      "rmdir",
    ],
  },
]
`;
